src/Offcanvas.php

```diff
@@ -4,235 +4,554 @@
+     *
+     * @see {\Yiisoft\Html\Html::renderTagAttributes()} for details on how attributes are being rendered.
+     */
+    public function titleAttributes(array $attributes): self
+    {
+        $new = clone $this;
+        $new->titleAttributes = $attributes;
+
+        return $new;
+    }
+
+    /**
+     * Sets the HTML attributes for the toggler.
+     *
+     * @param array $attributes Attribute values indexed by attribute names.
+     *
+     * @return self A new instance with the specified toggler attributes.
+     *
+     * @see {\Yiisoft\Html\Html::renderTagAttributes()} for details on how attributes are being rendered.
+     */
+    public function togglerAttributes(array $attributes): self
+    {
+        $new = clone $this;
+        $new->togglerAttributes = $attributes;
+
+        return $new;
+    }
+
+    /**
+     * Sets the content of the toggler.
+     *
+     * @param string|Stringable $content The content of the toggler.
+     *
+     * @return self A new instance with the specified toggler content.
+     */
+    public function togglerContent(string|Stringable $content): self
+    {
+        $new = clone $this;
+        $new->togglerContent = $content;
+
+        return $new;
+    }
+
+    /**
+     * Begins the rendering.
+     *
+     * @throws InvalidArgumentException if the tag is an empty string.
+     *
+     * @return string The opening HTML tags for the navbar.
+     */
+    public function begin(): string
+    {
+        parent::begin();
+
+        $id = $this->getId();
+        $html = '';
+
+        if ($this->togglerContent !== '' && $this->show === false) {
+            $html .= $this->renderToggler($id) . "\n";
         }
 
-        $html = $this->renderToggle ? $this->renderToggle() : '';
-        $html .= Html::openTag($tag, $options);
-        $html .= $this->renderHeader();
-        $html .= Html::openTag($bodyTag, $bodyOptions);
+        $html .= $this->renderOffcanvas($id);
 
         return $html;
     }
 
+    /**
+     * Run the widget.
+     *
+     * @return string The HTML representation of the element.
+     */
     public function render(): string
     {
-        $tag = $this->options['tag'] ?? 'div';
-        $bodyTag = $this->bodyOptions['tag'] ?? 'div';
+        $html = Html::closeTag('div') . "\n";
+        $html .= Html::closeTag('div');
 
-        return Html::closeTag($bodyTag) . Html::closeTag($tag);
+        return $html;
     }
 
     /**
-     * Renders offcanvas header.
+     * Generates the ID.
      *
-     * @return string the rendering header.
+     * @throws InvalidArgumentException if the ID is an empty string or `false`.
+     *
+     * @return string The generated ID.
+     */
+    private function getId(): string
+    {
+        return match ($this->id) {
+            true => Html::generateId(self::NAME . '-'),
+            '', false => throw new InvalidArgumentException('The "id" property must be specified.'),
```

Comment:
```suggestion
            '', false => throw new InvalidArgumentException('The "id" must be specified.'),
```

Since we use public method, it's not a good idea to mention property.

Reply:
done.